analytics/src/main/java/com/segment/analytics/internal/AnalyticsClient.java

```diff
@@ -227,22 +269,28 @@ public void run() {
               log.print(VERBOSE, "Flushing messages.");
             }
           } else {
+            messageIdSizeMap.put(message.messageId(), messageSizeInBytes(message));
             messages.add(message);
           }
 
           Boolean isBlockingSignal = message == FlushMessage.POISON || message == StopMessage.STOP;
           Boolean isOverflow = messages.size() >= size;
 
           if (!messages.isEmpty() && (isOverflow || isBlockingSignal)) {
-            Batch batch = Batch.create(CONTEXT, messages);
-            log.print(
-                VERBOSE,
-                "Batching %s message(s) into batch %s.",
-                messages.size(),
-                batch.sequence());
-            networkExecutor.submit(
-                BatchUploadTask.create(AnalyticsClient.this, batch, maximumRetries));
-            messages = new ArrayList<>();
+            while (!messages.isEmpty()) {
```

Comment:
Oh so the logic in this part works like this:

whatever messages we have in messages list. In the next line where we create the batch over in this method
public static Batch createBatch

we are checking the size that we stored in the map as we are adding messages to the batch

![image](https://user-images.githubusercontent.com/92827184/149876876-df5affb1-3a6f-4c61-a1df-862f66135fad.png)


Lets say we have the scenario that you described. we have 29 messages that together are less than 500kb but the 30th message goes over the limit. This method will create a batch with just the 29 messages, go out of the method, send the batch and then since we are still inside the loop and messages is not empty yet we will create one batch with just the last 30th message. This scenario is tested in AnalyticsClientTest.java if you want to check it out

Reply:
hey @jorgen25 thanks for explaining. that's actually the same as what I meant by "inner loop will continue and pack the 30th message alone in a batch". so there will be another batch contains only the 30th message, which causes an underflow issue (meaning a batch is sent when it can be waited for messages to be sent together). consider if this happens very often, that means we are making a lot of upload api calls for only a single message, which contradicts the meaning of "batch" and does not scale very well.

Reply:
yes. please get rid of the while loop. a batch with a single message is not a deal breaker in a client-side app, but in server-side it is, because we have much more traffic in server.

Reply:
yeah makes sense

Thanks @wenxi-zeng ill work on the agreed changes 👍